sdk/src/Core/Amazon.Runtime/_bcl/IClientConfig.bcl.cs

```diff
@@ -70,5 +70,16 @@ public partial interface IClientConfig
         /// </summary>
         WebProxy GetWebProxy();
 
+        /// <summary>
+        /// Returns a WebProxy instance to use for HTTPS connections if an
+        /// explicit web proxy hasn't been configured.
+        /// </summary>
+        WebProxy GetHttpsProxy();
+
+        /// <summary>
+        /// Returns a WebProxy instance to use for HTTP connections if an
+        /// explicit web proxy hasn't been configured.
+        /// </summary>
+        WebProxy GetHttpProxy();
```

Comment:
Is there a reason these shouldn't return IWebProxy and move all the common GetHttpsProxy/GetHttpProxy code to the non _bcl/netstandard specific partial classes? I think there may have been breaking change concerns but wouldn't it only be breaking if GetWebProxy was changed? I haven't traced all the code so it is possible this must return the utils type of WebProxy.

Reply:
Hey @boblodgett , frankly I'm not sure; I believe you are correct that in theory it could be done, but I haven't done all that tracing myself (and I'm certainly far from being an expert on the aws dotnet sdk).

If we did want to do this I think that'd be fine -- but I'd suggest maybe we do it in another follow up PR that can be rolled out separately? Keep this change focused on just the environment variables (so not only do we get our rollout which we desperately need! but the chance of it blowing up anyone goes down until we can do all that tracing).

Presumably it was done this way for a reason, and if we want to change it that's not really my call, but I'd want to do it as it's own distinct change. I wanna have as close to the original as possible.

Reply:
We wouldn't be able to change it in a follow-up for the new methods because it would change the return type from WebProxy to IWebProxy which would then be a breaking change. @normj what direction do you think we should go with this?

Reply:
Not sure if we still want to change this, but i've pushed it up without this change (keeping the same patterns we've had for now), if we do wanna change this just lemme know.

Reply:
Unfortunately, I don't think we can do it the way you've specified here. In BCL we end up passing the result of `GetHttpsProxy`/`GetHttpProxy` to
`AmazonSecurityTokenServiceClient` which ends up passing the proxy to `SAMLAuthenticationController` which has the following block of code:

```c#
#if NETSTANDARD
        public IWebProxy ProxySettings { get; private set; }
#else
        public WebProxy ProxySettings { get; private set; }
#endif
```

And a `IWebProxy` can't be assigned to a `WebProxy` if we're not using NETSTANDARD (this also shows up in quite a few other places like `IAuthenticationController` -- which the SAML controller passes the proxy to as well). And I think changing this makes it a breaking change if we want to return `IWebProxy`, and change all the argument types of the underlying values being passed around.

I do also want to mention that BCL doesn't have: `Amazon.Runtime.Internal.Util.WebProxy` defined, so the types between BCL, and not would end up being two different `WebProxy`'s. Overall I think we'd end up in a very similar place, and it'd make this change significantly larger/more likely to break as we have to go change a whole bunch of types.

Reply:
Thanks for the reminder about the different implementations of `WebProxy`. I remember now the original target of the SDK to support .NET Core was .NET Standard 1.3 which didn't have `WebProxy` just the interface `IWebProxy`. That triggered the difference in signature return types. We can't do anything about that till we do a major version bump. You can still do the request change moving the new methods into the base partial with some conditional if checks.

```
#if BCL
        private static WebProxy GetWebProxyWithCredentials(string value)
#else
        private static Amazon.Runtime.Internal.Util.WebProxy? GetWebProxyWithCredentials(string value)
#endif
        {
            if (!string.IsNullOrEmpty(value))
            {
                var asUri = new Uri(value);
#if BCL
                var parsedProxy =  new WebProxy(asUri);
#else
                var parsedProxy =  new Amazon.Runtime.Internal.Util.WebProxy(asUri);
#endif
                if (!string.IsNullOrEmpty(asUri.UserInfo))
                {
                    var userAndPass = asUri.UserInfo.Split(':');
                    parsedProxy.Credentials = new NetworkCredential(
                        userAndPass[0],
                        userAndPass.Length > 1 ? userAndPass[1] : string.Empty
                    );
                }
                return parsedProxy;
            }

            return null;
        }

        /// <summary>
        /// Returns a WebProxy instance to use for HTTPS connections if an
        /// explicit web proxy hasn't been configured.
        /// </summary>
#if BCL
        public WebProxy GetHttpsProxy()
#else
        public IWebProxy GetHttpsProxy()
#endif
        {
            var explicitProxy = GetWebProxy();
            if (explicitProxy != null)
            {
                return explicitProxy;
            }
            return ClientConfig.GetWebProxyWithCredentials(Environment.GetEnvironmentVariable("https_proxy"));
        }

        /// <summary>
        /// Returns a WebProxy instance to use for HTTP connections if an
        /// explicit web proxy hasn't been configured.
        /// </summary>
#if BCL
        public WebProxy GetHttpProxy()
#else
        public IWebProxy GetHttpProxy()
#endif
        {
            var explicitProxy = GetWebProxy();
            if (explicitProxy != null)
            {
                return explicitProxy;
            }
            return ClientConfig.GetWebProxyWithCredentials(Environment.GetEnvironmentVariable("http_proxy"));
        }
```

Since we want essentially the same behavior for both .NET Framework and .NET Core it is worth putting the logic in the shared location with the specific conditional checks as opposed to forking the logic into separate platform files where the implementation could drift over time.

Reply:
I've gone ahead, and pushed up this change! Marking as resolved for now.